src/base/common/util/ternarySearchTree.ts

```diff
@@ -0,0 +1,113 @@
+const enum Dir {
+    Left = -1,
+    Mid = 0,
+    Right = 1
+}
+
+export interface IKeyIterator<K> {
+
+    next(): IKeyIterator<K>;
+    hasNext(): boolean;
+    
+}
+
+export interface ITernarySearchTree<K, V> {
+    forStrings<T>(): TernarySearchTree<string, T>;
+    
+    /** 
+     * TODO:
+    */
+    clear(): void;
+
+    /**
+     * 
+     * @param values 
+     * @param keys 
+     */
+    fill(values: readonly [K, V][] | V, keys?: readonly K[]): void;
+
+    /**
+     * 
+     * @param value 
+     * @param key 
+     */
+    set(value: V, key: K): V | undefined;
+
+    /**
+     * 
+     * @param key 
+     */
+    get(key: K): V | undefined;
+
+    /**
+     * TODO: test has, cat and cata
+     * @param key 
+     */
+    has(key: K): boolean;
+
+    /**
+     * TODO:
+     * @param key 
+     */
+    delete(key: K): void;
+    
+}
+
+class TernarySearchTreeNode<K, V> {
+    height: number = 1;
+
+    // current segment of the key, assigned by key[pos]
+    segment!: string;
+
+    // the entire key
+    key: K | undefined;
+
+    value: V | undefined;
+    left: TernarySearchTreeNode<K, V> | undefined;
+    mid: TernarySearchTreeNode<K, V> | undefined;
+    right: TernarySearchTreeNode<K, V> | undefined;
+
+    // AVL rotate
+    public rotateLeft(): TernarySearchTreeNode<K, V> {
```

Comment:
This method has a precondition: the current `TernarySearchTreeNode` has a `right` node (`node.right` is not nullity). Consider writing a `@assert` to indicate that in the JSDoc of this method.

eg.
```ts
/**
 * @assert Requires `node.right` to be non-nullity.
 */
public rotateLeft(): TernarySearchTreeNode<K, V> {
        const tmp = this.right!;
        this.right = tmp.left;
        tmp.left = this;
        return tmp;
    }
```

Reply:
But I mean, since this data structure is quite simple, it is ok not to write JSDoc on those methods. Depends on you.